### PR TITLE
Remove `@gen_code` macro

### PR DESCRIPTION
**Pull request by fredrikekre**

### CODE REVIEW

**Comment by KristofferC** [comment]
Benchmarks? Just in case.

**Comment by fredrikekre** [comment]
hmmmmm

https://gist.github.com/fredrikekre/c5cb3987748621911362260b75a7b5f8

**Comment by fredrikekre** [comment]
Identical code is generated...

**Comment by KristofferC** [comment]
Surely not for det and inv? It even starts to allocate.

**Comment by fredrikekre** [comment]
PR:

``` jl
julia> A = rand(Tensor{2,3}); As = symmetric(A);

julia> @code_native det(As)
    .section    __TEXT,__text,regular,pure_instructions
Filename: math_ops.jl
    pushq   %rbp
    movq    %rsp, %rbp
Source line: 84
    movsd   8(%rdi), %xmm0          ## xmm0 = mem[0],zero
    movsd   16(%rdi), %xmm1         ## xmm1 = mem[0],zero
    movsd   24(%rdi), %xmm2         ## xmm2 = mem[0],zero
    movsd   32(%rdi), %xmm3         ## xmm3 = mem[0],zero
    movsd   40(%rdi), %xmm4         ## xmm4 = mem[0],zero
Source line: 85
    movapd  %xmm1, %xmm5
    mulsd   %xmm2, %xmm5
    mulsd   %xmm4, %xmm2
    mulsd   %xmm0, %xmm4
    movapd  %xmm1, %xmm6
    mulsd   %xmm3, %xmm6
    subsd   %xmm6, %xmm4
    mulsd   %xmm0, %xmm4
    mulsd   %xmm3, %xmm0
    mulsd   %xmm3, %xmm3
    subsd   %xmm3, %xmm2
    mulsd   (%rdi), %xmm2
    subsd   %xmm4, %xmm2
    subsd   %xmm5, %xmm0
    mulsd   %xmm1, %xmm0
    addsd   %xmm2, %xmm0
Source line: 69
    popq    %rbp
    retq
    nop

julia> @code_native det(A)
    .section    __TEXT,__text,regular,pure_instructions
Filename: math_ops.jl
    pushq   %rbp
    movq    %rsp, %rbp
Source line: 84
    movsd   8(%rdi), %xmm0          ## xmm0 = mem[0],zero
    movsd   16(%rdi), %xmm1         ## xmm1 = mem[0],zero
    movsd   32(%rdi), %xmm2         ## xmm2 = mem[0],zero
    movsd   40(%rdi), %xmm3         ## xmm3 = mem[0],zero
    movsd   56(%rdi), %xmm4         ## xmm4 = mem[0],zero
    movsd   64(%rdi), %xmm5         ## xmm5 = mem[0],zero
Source line: 85
    movapd  %xmm3, %xmm6
    mulsd   %xmm4, %xmm6
    mulsd   %xmm1, %xmm4
    mulsd   %xmm2, %xmm1
    mulsd   %xmm5, %xmm2
    subsd   %xmm6, %xmm2
    mulsd   (%rdi), %xmm2
    mulsd   %xmm0, %xmm5
    subsd   %xmm4, %xmm5
    mulsd   24(%rdi), %xmm5
    subsd   %xmm5, %xmm2
    mulsd   %xmm3, %xmm0
    subsd   %xmm1, %xmm0
    mulsd   48(%rdi), %xmm0
    addsd   %xmm2, %xmm0
Source line: 69
    popq    %rbp
    retq
    nopw    %cs:(%rax,%rax)
```

**Comment by fredrikekre** [comment]
master:

``` jl
julia> A = rand(Tensor{2,3}); As = symmetric(A);

julia> @code_native det(A)
    .section    __TEXT,__text,regular,pure_instructions
Filename: math_ops.jl
    pushq   %rbp
    movq    %rsp, %rbp
Source line: 66
    movsd   8(%rdi), %xmm0          ## xmm0 = mem[0],zero
    movsd   16(%rdi), %xmm1         ## xmm1 = mem[0],zero
    movsd   32(%rdi), %xmm2         ## xmm2 = mem[0],zero
    movsd   40(%rdi), %xmm3         ## xmm3 = mem[0],zero
    movsd   56(%rdi), %xmm4         ## xmm4 = mem[0],zero
    movsd   64(%rdi), %xmm5         ## xmm5 = mem[0],zero
    movapd  %xmm3, %xmm6
    mulsd   %xmm4, %xmm6
    mulsd   %xmm1, %xmm4
    mulsd   %xmm2, %xmm1
    mulsd   %xmm5, %xmm2
    subsd   %xmm6, %xmm2
    mulsd   (%rdi), %xmm2
    mulsd   %xmm0, %xmm5
    subsd   %xmm4, %xmm5
    mulsd   24(%rdi), %xmm5
    subsd   %xmm5, %xmm2
    mulsd   %xmm3, %xmm0
    subsd   %xmm1, %xmm0
    mulsd   48(%rdi), %xmm0
    addsd   %xmm2, %xmm0
    popq    %rbp
    retq
    nopw    %cs:(%rax,%rax)

julia> @code_native det(As)
    .section    __TEXT,__text,regular,pure_instructions
Filename: math_ops.jl
    pushq   %rbp
    movq    %rsp, %rbp
Source line: 66
    movsd   8(%rdi), %xmm0          ## xmm0 = mem[0],zero
    movsd   16(%rdi), %xmm1         ## xmm1 = mem[0],zero
    movsd   24(%rdi), %xmm2         ## xmm2 = mem[0],zero
    movsd   32(%rdi), %xmm3         ## xmm3 = mem[0],zero
    movsd   40(%rdi), %xmm4         ## xmm4 = mem[0],zero
    movapd  %xmm1, %xmm5
    mulsd   %xmm2, %xmm5
    mulsd   %xmm4, %xmm2
    mulsd   %xmm0, %xmm4
    movapd  %xmm1, %xmm6
    mulsd   %xmm3, %xmm6
    subsd   %xmm6, %xmm4
    mulsd   %xmm0, %xmm4
    mulsd   %xmm3, %xmm0
    mulsd   %xmm3, %xmm3
    subsd   %xmm3, %xmm2
    mulsd   (%rdi), %xmm2
    subsd   %xmm4, %xmm2
    subsd   %xmm5, %xmm0
    mulsd   %xmm1, %xmm0
    addsd   %xmm2, %xmm0
    popq    %rbp
    retq
    nop
```

**Comment by fredrikekre** [comment]
PR:

``` jl
julia> A = rand(Tensor{2,3}); As = symmetric(A);

julia> @benchmark det($A)
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     5.547 ns (0.00% GC)
  median time:      5.560 ns (0.00% GC)
  mean time:        5.564 ns (0.00% GC)
  maximum time:     14.446 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark det($As)
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     4.518 ns (0.00% GC)
  median time:      4.531 ns (0.00% GC)
  mean time:        4.643 ns (0.00% GC)
  maximum time:     57.394 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

**Comment by fredrikekre** [comment]
master:

``` jl
julia> @benchmark det($A)
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     5.196 ns (0.00% GC)
  median time:      5.216 ns (0.00% GC)
  mean time:        5.221 ns (0.00% GC)
  maximum time:     13.943 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark det($As)
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     4.519 ns (0.00% GC)
  median time:      5.056 ns (0.00% GC)
  mean time:        5.389 ns (0.00% GC)
  maximum time:     55.215 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

**Comment by fredrikekre** [comment]
Oh, its identical for 2 and 3 dimension... Its only for 1D that there is a problem, maybe I should have seen that :p

**Comment by fredrikekre** [comment]
https://gist.github.com/fredrikekre/d0c3ca3bd37e7b4b1fb4d9ca3855dc07

Modified in this PR:
| `["other","det - dim 1 - order 2 - Float32"]` | 0.84 (5%) :white_check_mark: | 1.00 (1%)  |
| `["other","det - dim 2 - order 2 - Float64"]` | 0.89 (5%) :white_check_mark: | 1.00 (1%)  |
| `["other","det - dim 3 - order 2 - Float64"]` | 0.94 (5%) :white_check_mark: | 1.00 (1%)  |

| `["other","inv - dim 2 - order 2 - Float32"]` | 0.87 (5%) :white_check_mark: | 1.00 (1%)  |
| `["other","inv - dim 2 - order 2sym - Float32"]` | 1.09 (5%) :x: | 1.00 (1%)  |

| `["other","norm - dim 2 - order 4sym - Float32"]` | 0.83 (5%) :white_check_mark: | 1.00 (1%)  |
| `["other","norm - dim 2 - order 4sym - Float64"]` | 0.90 (5%) :white_check_mark: | 1.00 (1%)  |

| `["other","trace - dim 1 - order 2 - Float32"]` | 0.84 (5%) :white_check_mark: | 1.00 (1%)  |
| `["other","trace - dim 1 - order 2 - Float64"]` | 0.84 (5%) :white_check_mark: | 1.00 (1%)  |

Seems OK.

File: src/math_ops.jl [hunk 1]
```diff
@@ -70,7 +70,7 @@ julia> det(A)
     if dim == 1
         return quote
             $(Expr(:meta, :inline))
-            get_data(t)[$idx(1,1)]
+            get_data(t)[$(idx(1,1))]
```
Comment (fredrikekre) [inline src/math_ops.jl:73]: nice brackets... I interpolated the function into the expression and not the index

Comment (KristofferC) [reply src/math_ops.jl:73]: Woops haha.

**Comment by KristofferC** [comment]
LGTM

**Comment by fredrikekre** [comment]
Good catch with the benchmarks! This would likely never have been found otherwise, since it passed tests, and no one ever inverts a `Tensor{2, 1}` :+1:

**Comment by KristofferC** [comment]
For later, we could just write out the explicit expressions for det and inv.

**Pull request merged**